include/alpaka/mem/BoundaryIter.hpp

```diff
@@ -0,0 +1,395 @@
+                Vec<BoundaryType, Dim>([](int) { return BoundaryType::LOWER; }),
+                lowerHaloSizes,
+                upperHaloSizes};
+        }
+
+        [[nodiscard]] ALPAKA_FN_HOST_ACC constexpr BoundaryDirectionIter<Dim, LowHaloVecType, UpHaloVecType> end()
+            const
+        {
+            return BoundaryDirectionIter<Dim, LowHaloVecType, UpHaloVecType>{
+                Vec<BoundaryType, Dim>([](int) { return BoundaryType::OOB; }),
+                lowerHaloSizes,
+                upperHaloSizes};
+        }
+
+        [[nodiscard]] ALPAKA_FN_HOST_ACC static consteval uint32_t length()
+        {
+            return internal::ipow(3, Dim);
+        }
+
+        [[nodiscard]] ALPAKA_FN_HOST_ACC static constexpr auto dim()
+        {
+            return Dim;
+        }
+
+    private:
+        LowHaloVecType const lowerHaloSizes;
+        UpHaloVecType const upperHaloSizes;
+    };
+
+    template<alpaka::concepts::Vector LowHaloVecType, alpaka::concepts::Vector UpHaloVecType>
+    ALPAKA_FN_HOST_ACC BoundaryDirectionsContainer(LowHaloVecType const& lowerHalos, UpHaloVecType const& upperHalos)
+        -> BoundaryDirectionsContainer<LowHaloVecType::dim(), LowHaloVecType, UpHaloVecType>;
+
+    /** @brief Construct and return a single boundary direction specifying the middle of a volume.
+     */
+    template<uint32_t Dims>
+    [[nodiscard]] constexpr auto makeCoreBoundaryDirection(
+        alpaka::concepts::Vector auto const& lowerHalos,
+        alpaka::concepts::Vector auto const& upperHalos)
+    {
+        return BoundaryDirection<Dims, ALPAKA_TYPEOF(lowerHalos), ALPAKA_TYPEOF(upperHalos)>{
+            fillCVec<BoundaryType, Dims, BoundaryType::MIDDLE>(),
+            lowerHalos,
+            upperHalos};
+    }
+
+    /** @brief Construct and return a single boundary direction specifying the middle of a volume with symmetric halos.
+     */
+    template<uint32_t Dims>
+    [[nodiscard]] constexpr auto makeCoreBoundaryDirection(alpaka::concepts::Vector auto const& halos)
+    {
+        return BoundaryDirection<Dims, ALPAKA_TYPEOF(halos), ALPAKA_TYPEOF(halos)>{
+            fillCVec<BoundaryType, Dims, BoundaryType::MIDDLE>(),
+            halos,
+            halos};
+    }
+
+    /**
+     * @brief Construct and return a single boundary direction specifying the middle of a volume with all halo sizes
+     * set to 1.
+     */
+    template<uint32_t Dims>
+    consteval auto makeCoreBoundaryDirection()
+    {
+        return makeCoreBoundaryDirection<Dims>(fillCVec<uint32_t, Dims, 1u>());
+    }
+
+    /** @brief Construct and return a boundary direction container. This container can be iterated over. See
+     * BoundaryDirectionsContainer.
+     * This constructor uses a default halo size of 1 everywhere.
+     *
+     * @tparam Dims The dimensionality of the container.
+     */
+    template<uint32_t Dims>
+    [[nodiscard]] ALPAKA_FN_HOST_ACC constexpr auto makeBoundaryDirIterator()
```

Comment:
Please do not use `ALPAKA_FN_HOST_ACC` if the compiler is not requiring it. 
The keyword `constexpr` is creating `__host__ __device__` functions too.